### PR TITLE
Array index typecheck

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
* Fix a missing type check on array index values.
* Add `__repr__` to abstract types to improve readability of error messages.

### How I did it
When an index value is not a literal integer, call to `validate_expected_type` and expect  `IntegerAbstractType`

### How to verify it
Run tests.

### Cute Animal Picture
image

### CODE REVIEW

**Comment by fubuloubu** [comment]
Add a test

File: vyper/context/types/abstract.py [hunk 1]
```diff
@@ -16,16 +16,26 @@ def compare_type(self, other):
             pass
         return isinstance(other, type(self))
 
+    def __repr__(self):
+        value = super().__repr__()
+        if value.startswith("<"):
```
Comment (fubuloubu) [inline vyper/context/types/abstract.py:21]: Why does this mean it gives meaningful output?

File: vyper/context/types/indexable/sequence.py [hunk 2]
```diff
@@ -69,8 +70,10 @@ def get_index_type(self, node):
         if isinstance(node, vy_ast.Int):
             if node.value < 0:
                 raise ArrayIndexException("Vyper does not support negative indexing", node)
-            if node.value < 0 or node.value >= self.length:
```
Comment (fubuloubu) [inline vyper/context/types/indexable/sequence.py:72]: lol

File: vyper/context/types/abstract.py (hunk 1 above)
Comment (iamdefinitelyahuman) [reply vyper/context/types/abstract.py:21]: the opposite actually.  a normal repr is `<ClassName object at 0x...>`, whereas the desired ones will be e.g. "string".  If it opens with `<`, we know it's not implemented and so we use `_description`.

Comment (fubuloubu) [reply vyper/context/types/abstract.py:21]: Ah, could it be `f"<{self.__class__.__name__}"` instead, so we know that it's a normal Python repr?

**Pull request merged**